### ISSUE TITLE
EF Core Select error with Guid.Empty

### ISSUE THREAD

**Issue by markoweb2** [dotnet/efcore#27039]
This simplest of queries will throw an error:

```C#
var q = (from s in context.WhateverTable select new { Id = Guid.Empty }).ToList();
```

The error thrown is:
```
System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Guid'.
   at Microsoft.Data.SqlClient.SqlBuffer.get_Guid()
   at Microsoft.Data.SqlClient.SqlDataReader.GetGuid(Int32 i)
   at lambda_method136(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at SunHQ.MainWindow.<Window_Loaded>b__12_1() in C:\Users\Marko\Documents\Projects\SunHQ\SunHQ\MainWindow.xaml.cs:line 124
```

The reason for the error is because the following SQL syntax is generated:

`SELECT ''00000000-0000-0000-0000-000000000000'' AS [Id] FROM [WhateverTable] AS [s]`

Notice the double single quotes.
Please fix.

EF Core version: 6.0.1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows 11
IDE: Visual Studio 2022

**Comment by ajcvickers** [comment]
@markoweb2 The query posted above generates the following SQL with 6.0.1 on SQL Server:

```
info: 12/21/2021 09:07:47.018 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DbCommand (6ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT '00000000-0000-0000-0000-000000000000' AS [Id]
      FROM [Blogs] AS [b]
```

Did you maybe simplify from your real query? Can you attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate?

**Comment by markoweb2** [comment]
@ajcvickers - This is the most barebone simplified SQL query that I could write, that would show the error. Ofcourse this is not a real query that I would use.
Creating a small runnable project is impossible, because I cannot package an SQL server that would host the database in the runnable project. (I have not tested this bug with a LocalDB or other SQL compact versions)

The setup where this bug occurs, is an SQL Server 2017 running on VM (WinServer 2019), database first approach.
My devolpment machine is Win 11, Visual Studio 2022, .net 6.0.1 (WPF c# app), EF Core 6.0.1.
I used the scaffolding method to create the models.
And then simply running this query against any tabel in the model will throw the error.
(Maybe there are culture settings involved that create the double single quotes, both my pc and the servers is set to Estonian culture, OS language = English US)

Is there a public facing Azure SQL database instance, that I could write a small sample app against? If I can reproduce the error, I could send the runnable app project.

The only other option I can think of, is creating a minimal database in SQL, create a small app against the database that would show the error.
Then package the app and the database mdf file (but you would have to host the mdf file yourself in some SQL server and change the connection string to match)

Or a heavy duty approach, I could create a virtual machine, with Win 11 + SQL Express + Visual Studio, that you could launch and then verify the issue. Would have to host this 20-40GB file somewhere to download though (I might have some options)...

**Comment by markoweb2** [comment]
@ajcvickers - yes, I get an error.
Here is my code listing:

```
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using System;
using System.Linq;

namespace ConsoleApp2
{
    public static class Your
    {
        public static string ConnectionString = @"Data Source=.\SQLSTUDIO;Database=Studio;Integrated Security=SSPI;";
    }

    public class Studio
    {
        public int Id { get; set; }
    }

    public class SomeDbContext : DbContext
    {
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
            => optionsBuilder
                .UseSqlServer(Your.ConnectionString)
                .LogTo(Console.WriteLine, LogLevel.Information)
                .EnableSensitiveDataLogging();

        public DbSet<Studio> Studios { get; set; }
    }

    class Program
    {
        public static void Main(string[] args)
        {
            try
            {
                using (var context = new SomeDbContext())
                {
                    var q = (from s in context.Studios select new { Id = Guid.Empty }).ToList();
                }

                Console.WriteLine("Hello World!");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
            }
        }
    }
}
```

And here is the proof:
![image](https://user-images.githubusercontent.com/95127525/146910015-3a0f7722-c8dc-4d66-8c04-09e95728c771.png)

**Comment by ajcvickers** [comment]
@markoweb2 Thanks.

@smitpatel Issue happens when iterating over results. SqlClient is returning the value as a string, but metedata is Guid, and hence GetGuid fails.

```
Unhandled exception. System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Guid'.
   at Microsoft.Data.SqlClient.SqlBuffer.get_Guid()
   at Microsoft.Data.SqlClient.SqlDataReader.GetGuid(Int32 i)
   at lambda_method21(Closure , QueryContext , DbDataReader , ResultContext , SingleQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.Main() in C:\local\code\AllTogetherNow\SixOh\Program.cs:line 91
```

**Comment by markoweb2** [comment]
@ajcvickers - I'll quickly add my 2 cents.
If you noticed then EF Core is logging this query (single set of single quotes around Guid):
`info: 12/21/2021 09:07:47.018 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DbCommand (6ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT '00000000-0000-0000-0000-000000000000' AS [Id]
      FROM [Blogs] AS [b]`

But if you look at the query that arrives to SQL server with SQL's profile viewer, you should see double single quotes:
`SELECT ''00000000-0000-0000-0000-000000000000'' AS [Id] FROM [Blogs] AS [b]`

If you now turn the SQL's result into pseudo c# code, you can understand why the error happens:
```
string good = "00000000-0000-0000-0000-000000000000"; // (regular string double quotes)
string bad = "'00000000-0000-0000-0000-000000000000'"; // (extra single quote inside the string double " quote - this would be the c# string representation of the result from SQL)
var g1 = new Guid(good); // this works
var g2 = new Guid(bad); // this should throw the same error that can't convert String to Guid
```

So the issue as I see it, EF Core correctly interprets the c# code and tries to generate SQL with only a single set of single quotes.
But somewhere in the query transport chain (EF Core's SqlClient driver?), the single set of single quotes are turned into a double set of single quotes, when sending the query to SQL server. Thus the query itself becomes wrong.
SQL server itself correctly responds, but because the incoming query was wrongly formatted, the result that EF Core is then trying to parse, ends up with the `string bad = ....` example. Thus the exception.

**Comment by roji** [comment]
@markoweb2 I don't see any double quoting issue - I can clearly see EF Core sending `SELECT '00000000-0000-0000-0000-000000000000' ...`. Also, the exception above supports this (`Unable to cast object of type 'System.String' to type 'System.Guid'`) - had there been a double quote, that would have been a syntax error instead. Check exactly how you're viewing your SQL, the double quoting may be a simple display issue with your viewer or similar.

**Comment by roji** [comment]
Clearing milestone to treat this as a duplicate of #24075; we have a general problem with constant projections and should probably treat it as a single higher-level problem.

**Comment by markoweb2** [comment]
@roji  - hmm, I don't understand why the result is different now.

I took this line straight from SQL Server Profiler with copy-paste:
`SELECT ''00000000-0000-0000-0000-000000000000'' AS [Id] FROM [WhateverTable] AS [s]`

Just changed the table name manually to "WhateverTable".
I re-ran my tests today against two different SQL machines and now indeed the profiler is showing single quotes. (the bug still remains, can't cast String to Guid).

I don't know, maybe there has been some .NET update in the interim that has changed this... (I did see some minor updates in Windows Update for .NET 5 and 6)

I decided to double check what LINQ to SQL does with the same query.
It actually sends this query to SQL:
`SELECT NULL AS [EMPTY] FROM [dbo].[WhateverTable] AS [t0]`

And this query:
`var q = (from s in context.WhateverTable select new { Id = Guid.Empty, s.Name }).ToList();`

Becomes:
`SELECT t0.[Name] FROM [dbo].[WhateverTable] AS [t0]`

So basically the constant is never sent to SQL, it is instead handled on the client/c# side by LINQ to SQL.

I agree, that this is a duplicate of #24075; hopefully your team can fix this soon. I see it has been nearly a year since the original issue was reported and it still hasn't been fixed :(
